asciidtype/asciidtype/src/dtype.c

```diff
@@ -0,0 +1,265 @@
+#include "dtype.h"
+
+#include "casts.h"
+
+PyTypeObject *ASCIIScalar_Type = NULL;
+
+static PyObject *
+get_value(PyObject *scalar)
+{
+    PyObject *ret_bytes = NULL;
+    PyTypeObject *scalar_type = Py_TYPE(scalar);
+    if (scalar_type == &PyUnicode_Type) {
+        // attempt to decode as ASCII
+        ret_bytes = PyUnicode_AsASCIIString(scalar);
+        if (ret_bytes == NULL) {
+            PyErr_SetString(
+                    PyExc_TypeError,
+                    "Can only store ASCII text in a ASCIIDType array.");
+        }
+    }
+    else if (scalar_type != ASCIIScalar_Type) {
+        PyErr_SetString(PyExc_TypeError,
+                        "Can only store ASCII text in a ASCIIDType array.");
+        return NULL;
+    }
+    else {
+        PyObject *value = PyObject_GetAttrString(scalar, "value");
+        if (value == NULL) {
+            return NULL;
+        }
+        ret_bytes = PyUnicode_AsASCIIString(value);
+        Py_DECREF(value);
+    }
+    return ret_bytes;
+}
+
+/*
+ * Internal helper to create new instances
+ */
+ASCIIDTypeObject *
+new_asciidtype_instance(PyObject *size)
+{
+    ASCIIDTypeObject *new = (ASCIIDTypeObject *)PyArrayDescr_Type.tp_new(
+            (PyTypeObject *)&ASCIIDType, NULL, NULL);
+    if (new == NULL) {
+        return NULL;
+    }
+    long size_l = PyLong_AsLong(size);
+    if (size_l == -1 && PyErr_Occurred()) {
+        return NULL;
+    }
+    new->size = size_l;
+    new->base.elsize = size_l * sizeof(char);
+    new->base.alignment = size_l *_Alignof(char);
+
+    return new;
+}
+
+/*
+ * This is used to determine the correct dtype to return when operations mix
+ * dtypes (I think?). For now just return the first one.
+ */
+static ASCIIDTypeObject *
+common_instance(ASCIIDTypeObject *dtype1, ASCIIDTypeObject *dtype2)
+{
+    if (!PyObject_RichCompareBool((PyObject *)dtype1, (PyObject *)dtype2,
+                                  Py_EQ)) {
```

Comment:
N.B.:  This is actually the same as checking whether your resolver returns `NO_CASTING` or `EQUIV_CASTING`.